### PR TITLE
New C++ API and updated sample code

### PR DESCRIPTION
**Pull request by hodgman**
Hi, this PR contains a small change to the SDK Builder, allowing different type names to be used for parameter passing (which is required as C++ has special syntax for passing objects by reference), and a **massive** update to the C++ SDK.

The previous C++ API:
- Passed strings by value (instead of by reference) causing a lot of unnecessary allocation/copying.
- - This patch to the SDK Builder solves this problem.
- Had a huge amount (megabytes) of dependencies.
- - I've allowed the new version to build with zero dependencies, but in that situation, it's the user's responsibility to send the HTTP requests themselves.
- - The user can also "plug in" any HTTP and JSON library to get the full experience of the previous C++ API.
- - - I've included a sample application that uses `cpprest`'s JSON parsing and HTTP abilities (same library that the previous C++ API depended on), and another version of the sample that uses `rapidjson` and `curl`.
- - In order to achieve zero dependencies, I've included a simple, low performance ~200 line-of-code JSON parser. The user can easily replace it with any other JSON parsing library. I wrote this new parser specifically to embed in the PhantasmaAPI.h header, but also published it separately here: https://github.com/hodgman/locjson
- - - There's also a sample application that uses this zero-dependency version.
- Did things that make it hard for game developers to use it.
- - Exceptions aren't supported on some game development platforms, so I've made them optional and opt-in.
- - Use of the standard library (std::string/std::vector) is discouraged in many game engines, so I've made them optional.
- - Some game development platforms may require you to use a very specific (platform-specific) networking or HTTP library, so I've made this dependency able to be supplied by the user.
- - Some C++ projects are built with single-byte characters, and some with two-byte (wide) characters... This new API can support either situation.

Basically, I've re-written it to the standard that I'd expect from any kind of middleware product that was designed with game-developers as the target customer :)

### CODE REVIEW

**Comment by hodgman** [comment]
Rebased my changes against master :)

**Comment by Relfos** [comment]
Thank you for your submission, lots of good stuff here, we're now reviewing it
